### ISSUE TITLE
auto-detect elm version per file, instead of based on cwd

### ISSUE THREAD

**Issue by avh4** [avh4/elm-format#561]
In 0.8.1, the elm-version is auto-detected based on the presence of `elm.json` or `elm-package.json` in the current working directory.

Proposal: instead of checking the current working directory, for each file check the file's folder and if not found traverse upwards until a package file is found.

It seems like this would be more flexible, and still do the right thing in all current cases?

**Comment by lassik** [comment]
See also: #523, #545

**Comment by lassik** [comment]
I *think* it would be the right thing to traverse directories upwards, and I'm under the impression that this is what Git and many other tools do too. But not 100% confident about all edge cases that may come up. I'd say, ask a few very experienced programmers if there are any problems with it, and if none of them come up with anything, just do it 😄

**Comment by lassik** [comment]
What to do in case of `--stdin` is a good question.

Editor plugins are easiest to implement using `--stdin` for multiple reasons. The main one is that people often format code in the editor before saving the file, in which case the file contents would not up be to date. In Emacs, I launch the `elm-format` subprocess so that its current directory (cwd) is the directory where the formatted `.elm` file resides, so looking up `elm.json`/`elm-package.json` based on the cwd would work fine. Not sure what is done by other editor plugins.

The best alternative might be to have a command line flag like this (copy-pasted from [here](https://github.com/Unibeautify/unibeautify-cli/issues/18#issue-335917742)):

> Something that's really useful for editor plugins is an option to say what filename the contents of stdin are coming from.
> 
> * [prettier](https://github.com/prettier/prettier): this option is called `--stdin-filepath`
> * [rufo](https://github.com/ruby-formatter/rufo): this option is called `--filename`
> 
> In my experience, it's easiest and most reliable to implement an editor plugin by not using disk files at all. Instead, just feed unformatted code to stdin and get the formatted code back from stdout. But even when using stdin/stdout, it's useful if we can know the filename of the code (the code usually comes from a disk file, but may have been modified in the text editor and the modifications have not been saved, so the disk file is different from the editor buffer that is sent to stdin).
> 
> The point of knowing the filename is that:
> 
> * Formatter error/warning messages can show a filename that is intuitive to the user.
> * Unibeautify can find `.unibeautifyrc` and possibly other project settings files based on the filename.

If `--stdin` is given, you could have an error message like this:

> I couldn't figure out what version of Elm you are using!
>
> With `--stdin`, you should use either `--elm-version` to specify the Elm version directly, or `--stdin-filepath` to let elm-format know what disk file or directory corresponds to the code being formatted. elm-format will then look for `elm-package.json` or `elm.json` in that directory (and its parent directories) to figure out whether you are using Elm 0.18 or Elm 0.19.

As a final twist in the tale, people should be able to format code in temp buffers that don't have a filename associated with them. In that case I don't think there is anything sensible that can be done apart from requiring a manual `--elm-version` flag along with `--stdin`. The concept of a current directory is dubious for a temp buffer.

**Comment by lassik** [comment]
Doing the version check separately for each file given on the command line would definitely be the right thing, since they may (in principle, at least) be from different Elm projects.

And to make sure that `--elm-version` can be given only once so people don't think they can do things like `elm-format --elm-version 0.18 foo.elm --elm-version 0.19 bar.elm`. The check currently seems to be in place (the error message is a bit misleading, `Invalid option '--elm-version'`, but that's probably ok).

**Comment by lassik** [comment]
(Moved comments about --stdin to a separate issue, #563)

**Comment by lassik** [comment]
Doing the version check separately for each file given on the command line would definitely be the right thing, since they may (in principle, at least) be from different Elm projects.

And to make sure that `--elm-version` can be given only once so people don't think they can do things like `elm-format --elm-version 0.18 foo.elm --elm-version 0.19 bar.elm`. The check currently seems to be in place (the error message is a bit misleading, `Invalid option '--elm-version'`, but that's probably ok).

**Comment by rlefevre** [comment]
Currently, the presence of `elm.json` or `elm-package.json` is used also for `--stdin`.

Should this feature remove this dectection for `--stdin`? 
If so `--elm-version` would be required for `--stdin` or `0.19` would be used by else.

**Comment by avh4** [comment]
I haven't heard anyone bring this up in a long time, I'm guessing because very few folks have both 0.18 and 0.19 projects anymore?  If anyone thinks this would still be useful, please chime in.
